packages/components/src/components/CodeEditor/fields.ts

```diff
@@ -1,53 +1,127 @@
-import { StateEffect, StateEffectType, StateField } from "@codemirror/state";
+import { Facet, StateEffect, StateField } from "@codemirror/state";
 import { EditorView } from "@codemirror/view";
-import { useEffect } from "react";
+import { ReactNode, useEffect, useState } from "react";
 
-import { SqValuePath } from "@quri/squiggle-lang";
+import { SqProject, SqValuePath } from "@quri/squiggle-lang";
 
 import { Simulation } from "../../lib/hooks/useSimulator.js";
 
-// Helper class for turning any React prop into a CodeMirror field.
-// How to use:
-// 1. Create a field once statically with `ReactiveStateField.define<T>(initialValue)`
-// 2. Use `myField.field` as a CodeMirror extension
-// 3. Call `myField.use(view, prop)` as a hook to update the field with an effect.
-class ReactiveStateField<Value> {
-  effect: StateEffectType<Value>;
-  field: StateField<Value>;
-
-  private constructor(initialValue: Value) {
-    this.effect = StateEffect.define<Value>();
-    this.field = StateField.define<Value>({
-      create: () => initialValue,
+export type CodemirrorReactProps = {
+  // Similar to React props for CodeEditor, but with null values; CodeMirror doesn't like `undefined` values.
+  // TODO - it should be possible to auto-generate this type with some clever TypeScript.
+  simulation: Simulation | null;
+  onFocusByPath: ((path: SqValuePath) => void) | null;
+  onChange: (value: string) => void;
+  onSubmit: (() => void) | null;
+  showGutter: boolean;
+  lineWrapping: boolean;
+  project: SqProject;
+  sourceId: string;
+  height: string | number | null;
+  renderImportTooltip:
+    | ((params: { project: SqProject; importId: string }) => ReactNode)
+    | null;
+};
+
+const defaultReactProps: CodemirrorReactProps = {
+  simulation: null,
+  onFocusByPath: null,
+  showGutter: false,
+  lineWrapping: true,
+  project: SqProject.create(),
+  sourceId: "fake",
+  onChange: () => {},
+  onSubmit: null,
+  height: null,
+  renderImportTooltip: null,
+};
+
+function makeReactPropFacet<T extends keyof CodemirrorReactProps>(field: T) {
+  return Facet.define<CodemirrorReactProps[T], CodemirrorReactProps[T]>({
+    combine: (value) => {
+      return value.length ? value[0] : defaultReactProps[field];
+    },
+  });
+}
+
+/**
+ * These facets are derived from the single
+ * [StateField](https://codemirror.net/docs/ref/#state.StateEffect) defined by
+ * `useReactPropsField` below.
+ *
+ * Alternatively, we could model them as separate fields. This would simplify
+ * some things (we don't really use aggregational capabilities of facets), but
+ * it would complicate other things (we'd have to call `useEffect` once per
+ * field). I'm not sure which approach is better.
```

Comment:
^ this comment explains another possible approach that would be interesting to explore.

There are some tradeoffs between two approaches, but even if we make a switch in the future, it's not going to be too bad (facets and fields mostly work in the same way in codemirror, so it won't affect the implementation of extensions).